cmd/goreleaser/internal/configure.go

```diff
@@ -40,243 +40,427 @@ const (
 	}
 }
 
-func DockerImages(dist string) []config.Docker {
-	var r []config.Docker
-	for _, arch := range Architectures {
-		if dist == K8sDistro && K8sDockerSkipArchs[arch] {
-			continue
-		}
-		switch arch {
-		case ArmArch:
-			for _, vers := range ArmVersions(dist) {
-				r = append(r, DockerImage(dist, arch, vers))
-			}
-		default:
-			r = append(r, DockerImage(dist, arch, ""))
+func newContainerImageManifests(dist, os string, archs []string) []config.DockerManifest {
+	tags := []string{`{{ .Version }}`, "latest"}
+	var r []config.DockerManifest
+	for _, imageRepo := range imageRepos {
+		for _, tag := range tags {
+			r = append(r, osDockerManifest(imageRepo, tag, dist, os, archs))
 		}
 	}
 	return r
 }
 
-// DockerImage configures goreleaser to build a container image.
-// https://goreleaser.com/customization/docker/
-func DockerImage(dist, arch, armVersion string) config.Docker {
-	dockerArchName := archName(arch, armVersion)
+type containerImageOptions struct {
+	armVersion string
+	winVersion string
+}
+
+func (o *containerImageOptions) version() string {
+	if o.armVersion != "" {
+		return o.armVersion
+	}
+	return o.winVersion
+}
+
+func newContainerImages(dist string, targetOS string, targetArchs []string, opts containerImageOptions) []config.Docker {
+	images := []config.Docker{}
+	for _, targetArch := range targetArchs {
+		images = append(images, dockerImageWithOS(dist, targetOS, targetArch, opts))
+	}
+	return images
+}
+
+type fullBuildConfig struct {
+	targetOS   string
+	targetArch []string
+	armVersion []string
+}
+
+func (c *fullBuildConfig) Build(dist string) config.Build {
+	buildConfig := config.Build{
+		ID:     dist + "-" + c.targetOS,
+		Dir:    "_build",
+		Binary: dist,
+		BuildDetails: config.BuildDetails{
+			Env:     []string{"CGO_ENABLED=0"},
+			Flags:   []string{"-trimpath"},
+			Ldflags: []string{"-s", "-w"},
+		},
+		Goos:   []string{c.targetOS},
+		Goarch: c.targetArch,
+		Goarm:  c.armVersion,
+	}
+	return buildConfig
+}
+
+func (c *fullBuildConfig) OS() string {
+	return c.targetOS
+}
+
+type preBuiltBuildConfig struct {
+	targetOS   string
+	targetArch []string
+	preBuilt   config.PreBuiltOptions
+}
+
+func (c *preBuiltBuildConfig) Build(dist string) config.Build {
+	return config.Build{
+		ID:       dist + "-" + c.targetOS,
+		Builder:  "prebuilt",
+		PreBuilt: c.preBuilt,
+		Dir:      "_build",
+		Binary:   dist,
+		Goos:     []string{c.targetOS},
+		Goarch:   c.targetArch,
+		Goarm:    ArmVersions(dist),
+	}
+}
+
+func (c *preBuiltBuildConfig) OS() string {
+	return c.targetOS
+}
+
+func dockerImageWithOS(dist, os, arch string, opts containerImageOptions) config.Docker {
+	osArch := osArch{os: os, arch: arch, version: opts.version()}
 	imageTemplates := make([]string, 0)
-	for _, prefix := range ImagePrefixes {
-		dockerArchTag := strings.ReplaceAll(dockerArchName, "/", "")
+	for _, prefix := range imageRepos {
+		dockerArchTag := strings.ReplaceAll(osArch.imageTag(), "/", "")
```

Comment:
couldn't this be done inside the imageTag method or are there cases where the `/` is needed?